zipline/utils/input_validation.py

```diff
@@ -30,6 +32,98 @@ def ensure_upper_case(func, argname, arg):
         )
 
 
+def ensure_dtype(func, argname, arg):
+    """
+    Argument preprocessor that converts the input into a numpy dtype.
+
+    Usage
+    -----
+    >>> import numpy as np
+    >>> from zipline.utils.preprocess import preprocess
+    >>> @preprocess(dtype=ensure_dtype)
+    ... def foo(dtype):
+    ...     return dtype
+    ...
+    >>> foo(float)
+    dtype('float64')
+    """
+    try:
+        return dtype(arg)
+    except TypeError:
+        raise TypeError(
+            "{func}() couldn't convert argument "
+            "{argname}={arg!r} to a numpy dtype.".format(
+                func=_qualified_name(func),
+                argname=argname,
+                arg=arg,
+            ),
+        )
+
+
+def expect_dtypes(*_pos, **named):
+    """
+    Preprocessing decorator that verifies inputs have expected numpy dtypes.
+
+    Usage
+    -----
+    >>> from numpy import dtype, arange
+    >>> @expect_dtypes(x=dtype(int))
+    ... def foo(x, y):
+    ...    return x, y
+    ...
+    >>> foo(arange(3), 'foo')
+    (array([0, 1, 2]), 'foo')
+    >>> foo(arange(3, dtype=float), 'foo')
+    Traceback (most recent call last):
+       ...
+    TypeError: foo() expected an argument with dtype 'int64' for argument 'x', but got dtype 'float64' instead.  # noqa
+    """
+    if _pos:
+        raise TypeError("expect_dtypes() only takes keyword arguments.")
```

Comment:
why do we accept *args just to raise? This will raise a type error without accepting `_pos`.